.github/workflows/nightly.yml

```diff
@@ -0,0 +1,229 @@
+  #          │ ┌───────────── hour (0 - 23)
+  #          │ │ ┌───────────── day of the month (1 - 31)
+  #          │ │ │ ┌───────────── month (1 - 12 or JAN-DEC)
+  #          │ │ │ │ ┌───────────── day of the week (0 - 6 or SUN-SAT)                                  
+  #          │ │ │ │ │
+  #          │ │ │ │ │
+  schedule: 
+    - cron: '0 0 * * *'  # basically running everyday at 12AM
+    # cron works with default branch (main) only: # https://github.community/t/on-schedule-per-branch/17525/2
+  
+  push:
+    branches:
+      # Because we can't schedule run for non-main branches,
+      # to ensure we are running the build on the staging branch, we can add push policy for it
+      - staging
+      # Development triggers, can be removed afterwards
+      - laserprec/nightly-ci
+
+# This file defines following CI workflow:
+#
+#  ┌──────────┐   ┌─────────┐   ┌────────┐
+#  │  static  ├─┬─►  build* ├─┬─►  test  │
+#  │ analysis │ │ │  (cpu)  │ │ │ report │
+#  └──────────┘ │ └─────────┘ │ └────────┘         
+#               │ ┌─────────┐ │
+#               ├─►  build* ├─┤
+#               │ │ (spark) │ │
+#               │ └─────────┘ │
+#               │ ┌─────────┐ │
+#               └─►  build* ├─┘
+#                 │  (gpu)  │
+#                 └─────────┘
+#                    ....
+# *each runs in PARALLEL the different combinations
+#  of python version, OS, test subsets, etc
+#
+#  ASCII chart created via https://asciiflow.com/
+
+jobs:
+###############################################
+############### STATIC-ANALYSIS ###############
+###############################################
+  static-analysis:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Use Python 3.6
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.6
+    
+    - name: Install dependencies (tox)
+      run: |
+        python -m pip install --upgrade pip setuptools wheel
+        pip install tox
+
+    - name: Run flake8
+      # TODO: re-enable this flake8 block (turned off to get a draft of the pipeline infrastructure)
+      continue-on-error: true
+      run: |
+       tox -e flake8
+
+###############################################
+################## CPU-BUILD ##################
+###############################################
+  build-cpu:
+    runs-on: ${{ matrix.os }}
+    needs: static-analysis
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        python: [3.6]
+        # different kind of tests are located in tests/<unit|integration|smoke> folders
+        test-kind: ['unit', 'smoke']
+        # pytest markers configured in tox.ini. See https://docs.pytest.org/en/6.2.x/example/markers.html
+        test-marker: ['not spark and not gpu']
+        include:
+          # Integration tests needs a more powermachine with more memory. GitHub-hosted agents only have 7GB memory.
+          - os: self-hosted
+            python: 3.6
+            test-kind: 'integration'
```

Comment:
one question about this, is this pipeline running unit, smoke and integration test?

Reply:
Yup! I know this extra `include` clause is looking unusual. At the moment, the integration tests cannot run on the smaller SKUs (7GB) provided by GitHub because there is enough memory to run the training on the full dataset. 

Ideally, all our tests shouldn't be that compute intensive and they can be ran on the default SKUs provided by the CI (unless we need to test for compatibility with GPU enabled machines). Thus the matrix should ideally look like:

```yaml
   test-kind: ['unit', 'smoke', 'integration']
   test-marker: [....]
   ....
```
I think we can have a discussion over the objective of the different unit/smoke/integration tests and whether doing training on the full dataset achieve those goals. Added a new issue for the discussion: #1507.

Reply:
sorry I don't understand, you mentioned that this pipeline runs unit, smoke and integration but later that the integration cannot run on the smaller SKUs. Some questions:
1) Where are the integration tests running? 
2) Is there a UI where I can see an example on how the tests are running?
3) there are 3 tests running for CPU, unit, smoke and integration, are they running in parallel or sequentially?
4) Do we know if GitHub has plans to support bigger SKUs?
5) Related to 4, Azure Pipelines also has SKUs for testing, are they the same as GitHub, or are they bigger? (if they are bigger, then GitHub might catch up eventually)

Reply:
No worries! 

1. The integration tests are currently running in the [self-hosted gpu agent](https://github.com/microsoft/recommenders/settings/actions/runners). This is configured with the `self-hosted` keyword.
  ```
  include:
          - os: self-hosted
  ```
2. Please find all of the GitHub Action runs [here](https://github.com/microsoft/recommenders/actions), especially the `pr-gate` and `nightly` pipeline.
3. Current all of them are running in parallel. However because we currently set up only 2 self-hosted but there are about 5 nightly step needing to run on self-hosted machine, we haven't scale up the resource to run them in true parallel.
4. As far as I know, I don't think there's a plan. Maybe the SKU offerings are different with GitHub Enterprise, but I haven't gotten too much info on that.
5. Azure Pipeline and GitHub Action both uses the same SKU: Standard_DS2_v2.

I hope that helps.

Reply:
got it thanks for the clarification
1. would you mind explaining in the file that some test are being run on the hosted machine and that some tests on the self-hosted one?
2. For nightly, in the current pipeline we are running only smoke and integration, see https://github.com/microsoft/recommenders/blob/main/tests/ci/azure_pipeline_test/dsvm_nightly_linux_cpu.yml#L30, and these two are sequential. The reason why they are sequential is that the smoke tests are just a small version of the integration tests. Is it possible to replicate this behavior in github?

We are making so many changes in test, I think we need also to review the test readme, there might be info that is not up to date: https://github.com/microsoft/recommenders/blob/main/tests/README.md

Reply:
Thanks for the great feedbacks! 

1. I already added a comment on these mixed runs for integration tests and link to a new issue (#1507) on whether we should optimize/trim these integration tests :
```yml
 include:
          # Integration tests need a powerful machine with more memory. GitHub-hosted agents only have 7GB memory.
          # TODO: Optimization/refactoring should be done to ensure that these test can run in the default CI agent (#1507)
          - os: self-hosted
```
@miguelgfierro is there anywhere else you would like me to comment?

2. If I understand correctly, in the existing pipeline, the smoke tests are running before integration tests so we can "fail fast" and avoid running the full integration test which is destined to fail. In the similar manner, github action has the `[fail-fast](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast)` keyword which will cancel running jobs in the matrix if any failed.  If this isn't what you are going after and we need to run both of them sequentially, I think we would need to refactor the CI structure because is designed to run everything in parallel as much as possible.

> We are making so many changes in test, I think we need also to review the test readme
This is a great document when I was onboarding :). All of the CI changes is compatible with the what's written in the document in terms of test execution. One thing that we can change is that we no longer need the `--duration 0` flag from 
`pytest tests/integration -m "integration and not spark and not gpu" --durations 0`, because these flags are stored in [tox.ini](https://github.com/microsoft/recommenders/blob/main/tox.ini#L53) and pytest will pick the flags up from tox.ini without you specifying it. Of course, it doesn't hurt to repeat yourself add them in when running. 😃